File: python/demos/demo_translate_sweep.py
"""demo_translate_sweep.py

Pre-tensor sliding-comparison sweep with ``translate_attributes`` and the
raw-MA list mode of ``cos_sim_exp_tens``.

Scenario: a 3-note motif (C E G) hidden inside a 7-note melody
(D E F C E G A, one note per second). The motif appears exactly at
reference times 3, 4, 5. At each (pitch transposition, time shift)
offset, the query is translated and compared to the un-shifted
reference with the closed-form cosine similarity of a 2-attribute
MAET (pitch periodic at the octave; time absolute non-periodic). The
sweep should peak at (0 cents, 3 s) where the query aligns with the
embedded C-E-G, and at (1200 cents, 3 s) by octave periodicity.

The workflow is two function calls: one to ``translate_attributes``, one
to ``cos_sim_exp_tens`` (raw-MA scalar-vs-list form, with the
translated ``p_attr`` list as one operand and the reference
``p_attr`` as the other). The build step is internalised: the
reference is built once, each translated query once. The sweep is
specified as a length-A offsets list, one row of M candidate shifts per
attribute; Section 3 builds it. Section 7 shows the same sweep as a
single call to ``sweep_cos_sim_exp_tens``, which never builds the M
translated queries at all.

Compare ``windowed_similarity`` (see ``demo_helix_blend`` and
``demo_tempo_invariance``), which windows the context by event
weighting before each build --- the window multiplies per-event weights
and the window axis is then marginalized --- so that locality is
decoupled from the query's own support. The route used here returns a
strict cosine similarity (bounded in [0, 1] for non-negative weights)
and does not require choosing a window family.

See also
--------
mpt.translate_attributes
mpt.cos_sim_exp_tens
mpt.sweep_cos_sim_exp_tens
mpt.build_exp_tens
mpt.windowed_similarity
"""

import numpy as np
import matplotlib.pyplot as plt

import mpt


# =====================================================================
# 1. Build the reference melody and the query motif (pre-tensor form)
# =====================================================================

print("=== 1. Pre-tensor inputs ===")

# Reference: D-E-F-C-E-G-A at one note per second. The query C-E-G
# appears exactly at times 3, 4, 5.
ref_midi  = np.array([62, 64, 65, 60, 64, 67, 69])
ref_pitch = mpt.transform_attributes(ref_midi, None, ('midi', 'cents')).reshape(1, -1)
ref_time  = np.arange(7, dtype=float).reshape(1, -1)
ref_pAttr = [ref_pitch, ref_time]

# Query: C-E-G triad, 1-second spacing.
qry_midi  = np.array([60, 64, 67])
qry_pitch = mpt.transform_attributes(qry_midi, None, ('midi', 'cents')).reshape(1, -1)
qry_time  = np.arange(3, dtype=float).reshape(1, -1)
qry_pAttr = [qry_pitch, qry_time]

# Per-attribute geometry: pitch (attribute 0) is periodic at the
# octave; time (attribute 1) is absolute non-periodic.
sigma   = [50.0, 0.3]
r       = [1, 1]
is_rel  = [False, False]
is_per  = [True,  False]
periods = [1200.0, 0.0]

print("  reference: D-E-F-C-E-G-A, one note per second")
print("  query    : C-E-G triad, 1-second spacing")
print("  (the motif appears exactly at reference times 3, 4, 5)")
print(f"  sigma    : {sigma[0]:.0f} cents (pitch) / {sigma[1]:.2f} s (time)")
print()


# =====================================================================
# 2. Construct the (pitch, time) offset sweep
# =====================================================================

print("=== 2. Offset sweep grid ===")

# Pitch offsets: 0-1200 cents in 100-cent steps (one octave). The
# expected peak at pitch shift 0 is also visible at 1200 cents because
# the pitch group is octave-periodic.
pitch_grid = np.arange(0, 1201, 100, dtype=float)
# Time offsets: -1 to 5 seconds in 0.25-second steps.
time_grid  = np.arange(-1.0, 5.001, 0.25)

P_mesh, T_mesh = np.meshgrid(pitch_grid, time_grid, indexing="ij")
M = P_mesh.size
# P_mesh and T_mesh are used in Section 3 to build the offsets list.

print(f"  pitch grid: {pitch_grid.size} transpositions over one octave "
      f"(100-cent steps)")
print(f"  time  grid: {time_grid.size} positions from t = "
      f"{time_grid.min():.1f} to t = {time_grid.max():.1f} s")
print(f"  total sweep positions: M = {M}")
print()


# =====================================================================
# 3. Pre-tensor translation: two equivalent offset forms
# =====================================================================

print("=== 3. translate_attributes (offset sweep) ===")

# offsets is a length-A list, one entry per attribute. Each entry here
# is a (1, M) row, which the orientation grammar reads as a per-sweep
# global shift: M candidate offsets broadcast across the attribute's
# values (trivial here, as each attribute is single-value, K_a = 1). The
# M sweep columns are shared across attributes, so column m of every
# entry together defines the m-th translated copy. Reads naturally as
# "sweep pitch by these values; sweep time by these values".
offsets = [P_mesh.reshape(1, -1),     # pitch shifts (attribute 0)
           T_mesh.reshape(1, -1)]     # time  shifts (attribute 1)
qry_pAttr_swept, _, _ = mpt.translate_attributes(qry_pAttr, None, offsets)

# The returned list of translated copies carries its offsets with it (a
# TranslatedSweep), so cos_sim_exp_tens below can recognise the sweep;
# see Section 7.
print(f"  qry_pAttr_swept: {type(qry_pAttr_swept).__name__}, "
      f"length {len(qry_pAttr_swept)}")
print(f"  each entry is a length-{len(qry_pAttr)} list of K_a x N "
      f"value matrices")
print()


# =====================================================================
# 4. Raw-MA scalar-vs-list cosine similarity: one call
# =====================================================================

print("=== 4. cos_sim_exp_tens (raw-MA list mode) ===")

S_flat = np.asarray(mpt.cos_sim_exp_tens(
    ref_pAttr, None, qry_pAttr_swept, None,
    sigma, r, is_rel, is_per, periods,
    verbose=False,
))
S = S_flat.reshape(P_mesh.shape)         # (pitch, time) heatmap

i, j = np.unravel_index(int(np.argmax(S)), S.shape)
print(f"  cosine similarity surface: shape {S.shape} (pitch x time)")
print(f"  max similarity {S.max():.4f} at pitch shift "
      f"{P_mesh[i, j]:.0f} c, time shift {T_mesh[i, j]:.2f} s")
print("  (expected: 0 cents, 3.00 s --- the embedded C-E-G)")
print()


# =====================================================================
# 5. Visualise the sweep
# =====================================================================

print("=== 5. Plot ===")

fig, ax = plt.subplots(figsize=(9, 6))
im = ax.imshow(
    S.T,                     # transpose: x-axis pitch, y-axis time
    origin="lower",
    aspect="auto",
    extent=[pitch_grid[0], pitch_grid[-1],
            time_grid[0], time_grid[-1]],
)
fig.colorbar(im, ax=ax, label="cosine similarity")
ax.set_xlabel("Pitch transposition (cents)")
ax.set_ylabel("Time shift (s)")
ax.set_title("Pre-tensor sliding-comparison: cosine similarity\n"
             "Reference: D-E-F-C-E-G-A; query: C-E-G")
ax.set_xticks(np.arange(0, 1201, 200))
ax.set_yticks(np.arange(-1, 5.01, 1))

# Mark the expected peak positions: query aligns with the embedded
# C-E-G at (0 c, 3 s). Octave periodicity reproduces the peak at
# (1200 c, 3 s).
ax.plot([0, 1200], [3, 3], "rx", markersize=12, mew=1.5)
ax.text(40,   3.4, "C-E-G match", color="r",
        fontsize=9, bbox=dict(facecolor="white", alpha=0.7,
                              edgecolor="none"))
ax.text(1100, 3.4, "octave", color="r",
        fontsize=9, bbox=dict(facecolor="white", alpha=0.7,
                              edgecolor="none"))

plt.tight_layout()
print("  Figure shows the cosine-similarity surface as a function of")
print("  pitch transposition and time shift. Red x marks the")
print("  expected peaks at (0 c, 3 s) and (1200 c, 3 s), where the")
print("  query aligns with the embedded C-E-G in the reference;")
print("  octave-pitch periodicity makes the two peaks identical.")
print()


# =====================================================================
# 6. Equivalent explicit build loop, for transparency
# =====================================================================

print("=== 6. Equivalent explicit build loop ===")
print("  This is what the raw-MA list mode does internally; spelled")
print("  out here so the relationship between translate_attributes,")
print("  build_exp_tens, and cos_sim_exp_tens is transparent.")
print()

dens_ref = mpt.build_exp_tens(
    ref_pAttr, None, sigma, r, is_rel, is_per, periods, verbose=False,
)
S_manual = np.empty(M, dtype=np.float64)
for m, pa in enumerate(qry_pAttr_swept):
    dens_q = mpt.build_exp_tens(
        pa, None, sigma, r, is_rel, is_per, periods, verbose=False,
    )
    S_manual[m] = mpt.cos_sim_exp_tens(dens_ref, dens_q, verbose=False)

discrepancy = float(np.max(np.abs(S_flat - S_manual)))
print(f"  max |S_raw - S_manual| = {discrepancy:.2e} "
      f"(floating-point parity)")
assert discrepancy < 1e-12, \
    "Raw-MA list mode disagrees with manual build loop."


# =====================================================================
# 7. The same sweep without building M queries: sweep_cos_sim_exp_tens
# =====================================================================

print("\n=== 7. sweep_cos_sim_exp_tens (one call, no translated copies) ===")
print("  A uniform translation of the query enters the inner product only")
print("  through the offset, so the whole sweep is one pass over the tuple")
print("  pairs and then one evaluation per offset. The pitch attribute is")
print("  periodic, which the mixture route refuses; under method='auto'")
print("  the orbit route carries the sweep instead (the wrapped kernel")
print("  absorbs the periodicity), so the call is the same either way.")

dens_qry = mpt.build_exp_tens(
    qry_pAttr, None, sigma, r, is_rel, is_per, periods, verbose=False,
)
offsets_am = np.vstack([P_mesh.ravel(), T_mesh.ravel()])      # (A, M)
S_sweep = mpt.sweep_cos_sim_exp_tens(dens_ref, dens_qry, offsets_am,
                                     verbose=False)
discrepancy_sweep = float(np.max(np.abs(S_flat - S_sweep)))
print(f"  max |S_raw - S_sweep| = {discrepancy_sweep:.2e}")
assert discrepancy_sweep < 1e-8, \
    "sweep_cos_sim_exp_tens disagrees with the per-offset route."

print("\n=== Demo complete ===")
plt.show()

File: matlab/demos/demo_translateSweep.m
%% demo_translateSweep.m 
% Pre-tensor sliding-comparison sweep with translateAttributes and the 
% raw-MA list mode of cosSimExpTens.
%
% Scenario: a 3-note motif (C E G) hidden inside a 7-note melody
% (D E F C E G A, one note per second). The motif appears exactly at
% reference times 3, 4, 5. At each (pitch transposition, time shift)
% offset, the query is translated and compared to the un-shifted
% reference with the closed-form cosine similarity of a 2-attribute
% MAET (pitch periodic at the octave; time absolute non-periodic).
% The sweep should peak at (0 cents, 3 s) where the query aligns
% with the embedded C-E-G, and at (1200 cents, 3 s) by octave
% periodicity.
%
% The workflow is two function calls: one to translateAttributes, one to
% cosSimExpTens (raw-MA scalar-vs-list form, with the translated p_attr
% list as one operand and the reference pAttr as the other). The build
% step is internalised: the reference is built once, each translated
% query once. The sweep is specified as a 1-by-A offsets cell, one row
% of M candidate shifts per attribute; Section 3 builds it. Section 7
% shows the same sweep as a single call to sweepCosSimExpTens, which
% never builds the M translated queries at all.
%
% Compare windowedSimilarity (see demo_helixBlend and
% demo_tempoInvariance), which windows the context by event weighting
% before each build --- the window multiplies per-event weights and the
% window axis is then marginalized --- so that locality is decoupled from
% the query's own support. The route used here returns a strict cosine
% similarity (bounded in [0, 1] for non-negative weights) and does not
% require choosing a window family.
%
% See also TRANSLATEATTRIBUTES, COSSIMEXPTENS, SWEEPCOSSIMEXPTENS,
% BUILDEXPTENS, WINDOWEDSIMILARITY.

clear; clc;

%% ===================================================================
%  1. Build the reference melody and the query motif (pre-tensor form)
%  ===================================================================

fprintf('=== 1. Pre-tensor inputs ===\n');

% Reference: D-E-F-C-E-G-A at one note per second. The query C-E-G
% appears exactly at times 3, 4, 5.
refMidi   = [62 64 65 60 64 67 69];
refPitch  = transformAttributes(refMidi, [], {'midi', 'cents'});
refTime   = 0:6;
refPAttr  = {refPitch, refTime};

% Query: C-E-G triad, 1-second spacing, sweep across both axes.
qryMidi   = [60 64 67];
qryPitch  = transformAttributes(qryMidi, [], {'midi', 'cents'});
qryTime   = 0:2;
qryPAttr  = {qryPitch, qryTime};

% Per-attribute geometry: pitch (attribute 1) is periodic at the
% octave; time (attribute 2) is absolute non-periodic.
sigma     = [50, 0.3];        % per-attribute sigma: cents, seconds
r         = [1, 1];           % single-value per attribute (K_a = 1)
isRel     = [false, false];
isPer     = [true,  false];
periods   = [1200,  0];

fprintf('  reference: D-E-F-C-E-G-A, one note per second\n');
fprintf('  query    : C-E-G triad, 1-second spacing\n');
fprintf('  (the motif appears exactly at reference times 3, 4, 5)\n');
fprintf('  sigma    : %.0f cents (pitch) / %.2f s (time)\n', ...
        sigma(1), sigma(2));
fprintf('\n');

%% ===================================================================
%  2. Construct the (pitch, time) offset sweep
%  ===================================================================

fprintf('=== 2. Offset sweep grid ===\n');

% Pitch offsets: 0 - 1200 cents in 100-cent steps (one octave). The
% expected peak at pitch shift 0 is also visible at 1200 cents because
% the pitch group is octave-periodic.
pitchGrid = 0:100:1200;
% Time offsets: -1 to 5 seconds in 0.25-second steps.
timeGrid  = -1:0.25:5;

[Pmesh, Tmesh] = meshgrid(pitchGrid, timeGrid);
M = numel(Pmesh);
% Pmesh and Tmesh are flattened in Section 3 into the per-attribute
% rows of the offsets cell.

fprintf('  pitch grid: %d transpositions over one octave (100-cent steps)\n', ...
        numel(pitchGrid));
fprintf('  time  grid: %d positions from t = %.1f to t = %.1f s\n', ...
        numel(timeGrid), min(timeGrid), max(timeGrid));
fprintf('  total sweep positions: M = %d\n', M);
fprintf('\n');

%% ===================================================================
%  3. Pre-tensor translation: build the swept query
%  ===================================================================

fprintf('=== 3. translateAttributes (offset sweep) ===\n');

% offsets is a 1-by-A cell, one entry per attribute. Each entry here is
% a 1-by-M row, which the orientation grammar reads as a per-sweep
% global shift: M candidate offsets broadcast across the attribute's
% values (trivial here, as each attribute is single-value, K_a = 1). The
% M sweep columns are shared across attributes, so column m of every
% entry together defines the m-th translated copy. Reads naturally as
% "sweep pitch by these values; sweep time by these values".
offsetsCell   = {Pmesh(:).', Tmesh(:).'};
[pmSwept, sweep] = translateAttributes(qryPAttr, [], offsetsCell);
qryPAttrSwept = pmSwept.pAttr;
% The fourth output records the per-attribute offsets (A x M) for
% sweepCosSimExpTens; see Section 7.

fprintf('  qryPAttrSwept: %s, length %d\n', class(qryPAttrSwept), ...
        numel(qryPAttrSwept));
fprintf('  each entry is a 1-by-%d cell of K_a-by-N value matrices\n', ...
        numel(qryPAttr));
fprintf('\n');

%% ===================================================================
%  4. Raw-MA scalar-vs-list cosine similarity: one call
%  ===================================================================

fprintf('=== 4. cosSimExpTens (raw-MA list mode) ===\n');

sCells = cosSimExpTens(refPAttr, [], qryPAttrSwept, [], ...
                        sigma, r, isRel, isPer, periods, ...
                        'verbose', false);
S      = cell2mat(sCells);             % 1-by-M
S      = reshape(S, size(Pmesh));      % size = [numel(timeGrid), numel(pitchGrid)]

% Locate the peak.
[sMax, iLin] = max(S(:));
[iT, iP]     = ind2sub(size(S), iLin);
fprintf('  cosine similarity profile: %d x %d (time x pitch)\n', ...
        size(S, 1), size(S, 2));
fprintf('  max similarity %.4f at pitch shift %.0f c, time shift %.2f s\n', ...
        sMax, pitchGrid(iP), timeGrid(iT));
fprintf('  (expected: 0 cents, 3.00 s --- the embedded C-E-G)\n');
fprintf('\n');

%% ===================================================================
%  5. Visualise the sweep
%  ===================================================================

fprintf('=== 5. Plot ===\n');

fig = figure('Name', 'demo\_translateSweep: pre-tensor sliding cosine', ...
             'Position', [100 100 900 600], 'Color', 'w');

imagesc(pitchGrid, timeGrid, S);
axis xy;
colorbar;
xlabel('Pitch transposition (cents)');
ylabel('Time shift (s)');
title({'Pre-tensor sliding-comparison: cosine similarity', ...
       'Reference: D-E-F-C-E-G-A; query: C-E-G'});
set(gca, 'XTick', 0:200:1200, 'YTick', -1:1:5);
hold on;

% Mark the expected peak positions: query aligns with the embedded
% C-E-G at (0 c, 3 s). Octave periodicity reproduces the peak at
% (1200 c, 3 s).
plot([0 1200], [3 3], 'rx', 'MarkerSize', 12, 'LineWidth', 1.5);
text(40,   3.4, 'C-E-G match', 'Color', 'r', ...
     'FontSize', 9, 'BackgroundColor', [1 1 1 0.7]);
text(1100, 3.4, 'octave', 'Color', 'r', ...
     'FontSize', 9, 'BackgroundColor', [1 1 1 0.7]);

fprintf('  Figure shows the cosine-similarity surface as a function of\n');
fprintf('  pitch transposition and time shift. Red x marks the\n');
fprintf('  expected peaks at (0 c, 3 s) and (1200 c, 3 s), where the\n');
fprintf('  query aligns with the embedded C-E-G in the reference;\n');
fprintf('  octave-pitch periodicity makes the two peaks identical.\n');
fprintf('\n');

%% ===================================================================
%  6. Equivalent explicit build loop, for transparency
%  ===================================================================

fprintf('=== 6. Equivalent explicit build loop ===\n');
fprintf('  This is what the raw-MA list mode does internally; here it\n');
fprintf('  is spelled out so the relationship between translateAttributes,\n');
fprintf('  buildExpTens, and cosSimExpTens is transparent.\n\n');

showPreMaet(refPAttr, [], [], 'names', {'pitch', 'time'}, ...
    'sigma', sigma, 'isRel', isRel, 'isPer', isPer, 'period', periods);
showPreMaet(qryPAttrSwept{1}, [], [], 'names', {'pitch', 'time'}, ...
    'sigma', sigma, 'isRel', isRel, 'isPer', isPer, 'period', periods);
fprintf('\n');

densRef = buildExpTens(refPAttr, [], sigma, r, ...
                       isRel, isPer, periods, 'verbose', false);
S_manual = zeros(1, M);
for m = 1:M
    densQ = buildExpTens(qryPAttrSwept{m}, [], sigma, r, ...
                         isRel, isPer, periods, 'verbose', false);
    S_manual(m) = cosSimExpTens(densRef, densQ, 'verbose', false);
end
S_manual = reshape(S_manual, size(Pmesh));

discrepancy = max(abs(S(:) - S_manual(:)));
fprintf('  max |S_raw - S_manual| = %.2e (floating-point parity)\n', ...
        discrepancy);
assert(discrepancy < 1e-12, 'Raw-MA list mode disagrees with manual build loop.');

%% ===================================================================
%  7. The same sweep without building M queries: sweepCosSimExpTens
%  ===================================================================

fprintf('\n=== 7. sweepCosSimExpTens (one call, no translated copies) ===\n');
fprintf('  A uniform translation of the query enters the inner product only\n');
fprintf('  through the offset, so the whole sweep is one pass over the tuple\n');
fprintf('  pairs and then one evaluation per offset. The pitch attribute is\n');
fprintf('  periodic, which the mixture route refuses; under ''method'', ''auto''\n');
fprintf('  the orbit route carries the sweep instead (the wrapped kernel\n');
fprintf('  absorbs the periodicity), so the call is the same either way.\n');

densQry = buildExpTens(qryPAttr, [], sigma, r, ...
                       isRel, isPer, periods, 'verbose', false);
S_sweep = sweepCosSimExpTens(densRef, densQry, sweep.offsets, 'verbose', false);
discrepancySweep = max(abs(S(:).' - S_sweep(:).'));
fprintf('  max |S_raw - S_sweep| = %.2e\n', discrepancySweep);
assert(discrepancySweep < 1e-8, 'sweepCosSimExpTens disagrees with the per-offset route.');

fprintf('\n=== Demo complete ===\n');
